### PR TITLE
Clang format

### PR DESCRIPTION
**Pull request by cosunae**
I think we need a common formatting as there is different people contributing to this project. 
So I am adding this PR with the hope that you like the idea of using clang-format. 
I find it quite convenient because it can be used from almost any editor, IDE, and it has tons of configuration possibilities. Me personally I dont mind which style, as far as we have a common one. 

It only solves the problem for C++, but that is already a big chunk of code here. 

We can integrate in our editors/IDEs and/or 
http://clang.llvm.org/docs/ClangFormat.html

### CODE REVIEW

**Comment by pspoerri** [comment]
How do we use it?

**Comment by cosunae** [comment]
With Qt creator you have to install the beautifier.
You only need to activate the beautifier and have this .clang-format in the root directory together with the CMakeLists.txt, where I put the file. 
You can configure the keyboard shortcut to trigger the format of a file or selected section. 

http://doc.qt.io/qtcreator/creator-beautifier.html

With vim, emacs there are instructions in the link I posted above.

**Comment by cosunae** [comment]
you can also use it in a script fashion. 
clang provides a clang-format binary that returns the formatted text according to the conf provided

**Comment by pspoerri** [comment]
Ah, great! I got it to run. 

Can we have:
- Braces around `for` loops. This should prevent bugs where incorrectly indented lines can be mistaken for a bug. 
- Newlines for the braces, because this will make the braces more clear. 
- `()` around multiple statements with logicals? Something like

```
bool DataFieldInfo::operator==(const DataFieldInfo &other) const {
    // Compare the least expensive elements first
    return (bytesPerElement_ == other.bytesPerElement_ && rank_ == other.rank_ &&

            iSize_ == other.iSize_ && jSize_ == other.jSize_ && kSize_ == other.kSize_ && lSize_ == other.lSize_ &&

            iMinusHalo_ == other.iMinusHalo_ && iPlusHalo_ == other.iPlusHalo_ && jMinusHalo_ == other.jMinusHalo_ &&
            jPlusHalo_ == other.jPlusHalo_ && kMinusHalo_ == other.kMinusHalo_ && kPlusHalo_ == other.kPlusHalo_ &&
            lMinusHalo_ == other.lMinusHalo_ && lPlusHalo_ == other.lPlusHalo_ &&

            name_ == other.name_ && type_ == other.type_ && metainfo_ == other.metainfo_);
}
```

 (from `DataFieldsInfo.cpp`) Is really really hard to parse.

**Comment by clementval** [comment]
agree to use it ... we should find a common style.

**Comment by cosunae** [comment]
I am not sure introducing {} or () can be done with a formatter, but I will try to look at that. 
This could probably more for clang-tidy. 

But please if you have preferences just commit this branch or create another PR with changes, as said, I dont care much which format or style as far as we have one
